software/common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2020-2022, RespiraWorks
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+
+#include "rx_buffer.h"
+
+/**
+ * FrameDetector detects frames of this format:
+ * ...MbbbbbbbbbbbM...
+ * Where M is frame MARK byte and b is any other byte. It is responsibility of sender to ensure that
+ * MARK characters appear in the incoming stream to only mark the start and the end of the frame.
+ *
+ * Correct stream of frames will look like this:
+ * ...MbbbbbbMMbbbbbbbbMMbbbbbbbbbMMbbbbbM...
+ * In practical terms, this is achieved by passing the stream through an EscapeStream object
+ * before sending it on the line (on the sender's end).
+ *
+ * RxBuffer is responsible for underlying reception and buffering of incoming data. Also RxBuffer
+ * provides callbacks described by RxListener interface. In particular:
+ *
+ *   on_character_match is called when MARK character is received
+ *   on_rx_complete is called when the requested number of bytes was received
+ *   on_rx_error is called when an error occurs in the underlying reception infrastructure
+ *
+ * FrameDetector is modeled as a Finite State Machine having states:
+ *
+ *   Lost - FrameDetector is lost in the stream of incoming bytes and will interpret the next MARK
+ *          byte as an end of the frame.
+ *   WaitForStartMarker - FrameDetector is synced and is waiting for MARK byte that will denote the
+ *          start of the frame
+ *   ReceivingFrame - FrameDetector is synced and will interpret the next MARK byte as an end of the
+ *          frame and will interpret the bytes in RxBuffer as a frame.
+ *
+ * on_rx_error will cause FrameDetector to transition to the Lost state, frame being received will
+ *      be ignored:
+ *          ...MbbbbbbEbbbbMMbbbbbbbM
```

Comment:
Are the M's identical? If so what happens if we start in the lost state an the first character we get 
is the second M? Will that be treated as end-of-frame?

I'm sure you have already covered this in the unit tests, but just want to ask anyway as framing is difficult.